File: app/src/main/java/com/kacper/and_krakgo/screens/sign_in/register/RegisterFragment.java
package com.kacper.and_krakgo.screens.sign_in.register;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v7.widget.AppCompatCheckBox;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.kacper.and_krakgo.InputTypes;
import com.kacper.and_krakgo.R;
import com.kacper.and_krakgo.helpers.FragmentHelper;
import com.kacper.and_krakgo.helpers.ToastMessageHelper;
import com.kacper.and_krakgo.helpers.ValidationHelper;
import com.kacper.and_krakgo.rxbus.SignInEvents;
import com.kacper.and_krakgo.screens.sign_in.register_part2.RegisterPart2Fragment;
import com.kacper.and_krakgo.views.CustomTextWatcher;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.functions.Consumer;

/**
 * Created by kacper on 22/10/2017.
 */

public class RegisterFragment extends Fragment implements RegisterContract.View {
    public static final String TAG = "RegisterFragment";

    @BindView(R.id.register_email_input) TextInputLayout mEmailInput;
    @BindView(R.id.register_password_input) TextInputLayout mPasswordInput;
    @BindView(R.id.register_retype_password_input) TextInputLayout mRetypePasswordInput;
    @BindView(R.id.above_18_checkbox) AppCompatCheckBox mAbove18CheckBox;
    @BindView(R.id.terms_accepted_checkbox) AppCompatCheckBox mtermsAcceptedCheckbox;
    @BindView(R.id.progress_bar) ProgressBar mProgressBar;

    private RegisterPresenter mPresenter;
    private boolean isBottomButtonEnabled = true;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_register_1, container, false);
        ButterKnife.bind(this, rootView);
        setListeners();
        return rootView;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPresenter = new RegisterPresenter(this);
        SignInEvents.subscribe(new Consumer<Integer>() {
            @Override
            public void accept(Integer integer) throws Exception {
                if (isVisible() && integer == SignInEvents.ACTION_BOTTOM_BUTTON_PRESSED)
                    goToNextPage();
            }
        });
    }

    private void goToNextPage() {
        if (!(mEmailInput.getEditText().length() > 0 && mPasswordInput.getEditText().length() > 0 && mRetypePasswordInput.getEditText().length() > 0) ||
                mEmailInput.isErrorEnabled() || mPasswordInput.isErrorEnabled() || mRetypePasswordInput.isErrorEnabled()) {
            ToastMessageHelper.showShortToast(R.string.error_fields_empty_or_invalid);
        } else if (!mAbove18CheckBox.isChecked()) {
            ToastMessageHelper.showShortToast(R.string.you_have_to_be_above_18);
        } else if (!mtermsAcceptedCheckbox.isChecked()) {
            ToastMessageHelper.showShortToast(R.string.you_have_to_accept_terms);
        } else if (isBottomButtonEnabled) {
            isBottomButtonEnabled = false;
            showProgress();
            mPresenter.signInWithEmail(mEmailInput.getEditText().getText().toString(),
                    mPasswordInput.getEditText().getText().toString());
        }
    }

    private void showProgress() {
        mEmailInput.setEnabled(isBottomButtonEnabled);
        mPasswordInput.setEnabled(isBottomButtonEnabled);
        mRetypePasswordInput.setEnabled(isBottomButtonEnabled);
        mAbove18CheckBox.setEnabled(isBottomButtonEnabled);
        mtermsAcceptedCheckbox.setEnabled(isBottomButtonEnabled);
        if (isBottomButtonEnabled)
            mProgressBar.setVisibility(View.GONE);
        else
            mProgressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void onSignUpCompleted() {
        FragmentHelper.removeAndChangeFragment(getFragmentManager(), new RegisterPart2Fragment());
    }

    @Override
    public void showError(Exception error) {
        if (error != null) {
            ToastMessageHelper.showShortToast(error.toString());
        }
        isBottomButtonEnabled = true;
        showProgress();
    }

    private void setListeners() {
        mEmailInput.getEditText().addTextChangedListener(new CustomTextWatcher() {
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                super.onTextChanged(charSequence, i, i1, i2);
                showOrHideError(charSequence.toString(), InputTypes.EMAIL);
            }
        });
        mPasswordInput.getEditText().addTextChangedListener(new CustomTextWatcher() {
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                super.onTextChanged(charSequence, i, i1, i2);
                showOrHideError(charSequence.toString(), InputTypes.PASSWORD);
            }
        });
        mRetypePasswordInput.getEditText().addTextChangedListener(new CustomTextWatcher() {
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                super.onTextChanged(charSequence, i, i1, i2);
                showOrHideError(charSequence.toString(), InputTypes.RETYPEPASSWORD);
                if (!mRetypePasswordInput.isErrorEnabled()) {
                    if (!mPasswordInput.getEditText().getText().toString().equals(charSequence.toString())) {
                        mRetypePasswordInput.setError(getString(R.string.password_dont_match));
                        mRetypePasswordInput.setErrorEnabled(true);
                    } else
                        mRetypePasswordInput.setErrorEnabled(false);
                }
            }
        });
    }

    private TextInputLayout getInputLayoutByType(InputTypes inputType) {
        switch (inputType) {
            case PASSWORD:
                return mPasswordInput;
            case RETYPEPASSWORD:
                return mRetypePasswordInput;
            default:
            case EMAIL:
                return mEmailInput;
        }
    }

    public void showOrHideError(String message, InputTypes type) {
        TextInputLayout textInputLayout = getInputLayoutByType(type);
        String error = ValidationHelper.validateText(getContext(), message, type);
        if (error == null) {
            textInputLayout.setErrorEnabled(false);
        } else {
            textInputLayout.setEnabled(true);
            textInputLayout.setError(error);
        }
    }

}

File: app/src/main/java/com/kacper/krakgo/screens/sign_in/register/RegisterFragment.kt
package com.kacper.krakgo.screens.sign_in.register

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.TextInputLayout
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kacper.krakgo.model.enums.InputTypes
import com.kacper.krakgo.R
import com.kacper.krakgo.helpers.FragmentHelper
import com.kacper.krakgo.helpers.ToastMessageHelper
import com.kacper.krakgo.helpers.ValidationHelper
import com.kacper.krakgo.rxbus.SignInEvents
import com.kacper.krakgo.screens.sign_in.register_part2.RegisterPart2Fragment
import com.kacper.krakgo.views.CustomTextWatcher

import butterknife.ButterKnife
import butterknife.OnClick
import com.kacper.krakgo.mvp.MvpFragment
import kotlinx.android.synthetic.main.fragment_register_1.*

/**
 * Created by kacper on 22/10/2017.
 */

class RegisterFragment : MvpFragment<RegisterContract.View, RegisterContract.Presenter>(),
        RegisterContract.View {

    override var mPresenter: RegisterContract.Presenter = RegisterPresenter()
    private var isBottomButtonEnabled = true

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_register_1, container, false)
        ButterKnife.bind(this, rootView)

        return rootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setListeners()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        SignInEvents.subscribe { integer ->
            if (isVisible && integer == SignInEvents.ACTION_BOTTOM_BUTTON_REGISTER_1)
                goToNextPage()
        }
    }

    private fun goToNextPage() {
        if (!(register_email_input.editText!!.length() > 0
                        && register_password_input.editText!!.length() > 0
                        && register_retype_password_input.editText!!.length() > 0) ||
                register_email_input.isErrorEnabled || register_password_input.isErrorEnabled
                || register_retype_password_input.isErrorEnabled) {
            ToastMessageHelper.showShortToast(context!!, R.string.error_fields_empty_or_invalid)
        } else if (!terms_accepted_checkbox.isChecked) {
            ToastMessageHelper.showShortToast(context!!, R.string.you_have_to_accept_terms)
        } else if (isBottomButtonEnabled) {
            isBottomButtonEnabled = false
            showProgress()
            mPresenter.signInWithEmail(register_email_input.editText!!.text.toString(),
                    register_password_input.editText!!.text.toString())
        }
    }

    private fun showProgress() {
        register_email_input.isEnabled = isBottomButtonEnabled
        register_password_input.isEnabled = isBottomButtonEnabled
        register_retype_password_input.isEnabled = isBottomButtonEnabled
        terms_accepted_checkbox.isEnabled = isBottomButtonEnabled
        if (isBottomButtonEnabled)
            progress_bar.visibility = View.GONE
        else
            progress_bar.visibility = View.VISIBLE
    }

    override fun onSignUpCompleted() {
        FragmentHelper.removeAndChangeFragment(fragmentManager!!, RegisterPart2Fragment())
    }

    override fun showMessage(messageId: Int) {

    }

    override fun showError(error: String) {
        ToastMessageHelper.showShortToast(context, error)
        isBottomButtonEnabled = true
        showProgress()
    }

    private fun setListeners() {
        register_email_input.editText!!.addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                super.onTextChanged(charSequence, i, i1, i2)
                showOrHideError(charSequence.toString(), InputTypes.EMAIL)
            }
        })
        register_password_input.editText!!.addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                super.onTextChanged(charSequence, i, i1, i2)
                showOrHideError(charSequence.toString(), InputTypes.PASSWORD)
            }
        })
        register_retype_password_input.editText!!.addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                super.onTextChanged(charSequence, i, i1, i2)
                showOrHideError(charSequence.toString(), InputTypes.RETYPEPASSWORD)
                if (register_retype_password_input.isErrorEnabled) {
                    if (register_password_input.editText!!.text.toString()
                            != charSequence.toString()) {
                        register_retype_password_input.error = getString(R.string.password_dont_match)
                        register_retype_password_input.isErrorEnabled = true
                    } else
                        register_retype_password_input.isErrorEnabled = false
                }
            }
        })
        terms_textview.setOnClickListener({
            val browserIntent = Intent(Intent.ACTION_VIEW,
                    Uri.parse("http://www.kacperkogut.com/pubgo/terms"))
            startActivity(browserIntent)
        })
    }

    private fun getInputLayoutByType(inputType: InputTypes): TextInputLayout? {
        return when (inputType) {
            InputTypes.PASSWORD -> register_password_input
            InputTypes.RETYPEPASSWORD -> register_retype_password_input
            InputTypes.EMAIL -> register_email_input
            else -> register_email_input
        }
    }

    fun showOrHideError(message: String, type: InputTypes) {
        val textInputLayout = getInputLayoutByType(type)
        val error = ValidationHelper.validateText(context!!, message, type)
        if (error == null) {
            textInputLayout!!.isErrorEnabled = false
        } else {
            textInputLayout!!.isEnabled = true
            textInputLayout.error = error
        }
    }

}
